### PR TITLE
Update router to allow async functions

### PR DESCRIPTION
**Pull request by felorhik**
This update is just to allow the router to be generated by an asynchronous function. A fairly common requirement. This would mainly be needed for special implementations of the router module, but since the router is built so early in the lifecycle it makes it hard to route properly if you need the API to define the routes for the app in some way (like a wordpress backend).

https://github.com/nuxt-community/router-module/issues/4

My implementation may be improved as I am not familiar enough with the codebase to be confident in it, the update is meant to open up the discussion on it. To me this update is extremely simple and should not cause any problems in nuxt, at least I have not seen any errors associated with the update in my own testing (I edited directly out of node_modules then applied the update to this fork, since this functionality is a requirement in the project I am working on).

Thoughts? The idea here does not need to be merged, more just to show a simple way to allow API driven routing.

Here is an example of, if using the router module how you can use the `router.js` file with this update.

```
export async function createRouter(options) {
    // since the axios module has not been initialized yet it needs to be included manually
    const response = await axios.get(process.env.API_URL + '/wp/v2/menus')
    console.log(response.data)
    const routes = []
    //loop response and build route array
    return new Router({
        mode: 'history',
        routes: routes
    })
}
```

### CODE REVIEW

File: lib/app/index.js [hunk 1]
```diff
@@ -44,7 +44,11 @@ const defaultTransition = <%=
 %>
 
 async function createApp (ssrContext) {
-  const router = createRouter(ssrContext)
+  if (createRouter.constructor.name === 'AsyncFunction') {
+    const router = await createRouter(ssrContext)
```
Comment (clarkdo) [inline lib/app/index.js]: router need to be defined outside the `if-else`

**Review by clarkdo — CHANGES_REQUESTED** [review]

File: lib/app/index.js (hunk 1 above)
Comment (clarkdo) [reply lib/app/index.js]: And how about to also consider returning `promise` ?

File: lib/app/index.js [hunk 2]
```diff
@@ -44,7 +44,7 @@ const defaultTransition = <%=
 %>
 
 async function createApp (ssrContext) {
-  const router = createRouter(ssrContext)
+  const router = await createRouter(ssrContext)
```
Comment (clarkdo) [inline lib/app/index.js]: If the value of the expression following the await operator is not a Promise, it's converted to a resolved Promise.

**Review by clarkdo — APPROVED** [review]

**Comment by clarkdo** [comment]
@jordanramstad Thanks

**Comment by jordanramstad** [comment]
Thanks, looks like I was overthinking how await would resolve, it was even simpler then I thought :)

**Pull request merged**